protected.go

```diff
@@ -135,13 +135,44 @@ func (p *Protector) resolve(op ops.Op, network string, addr string) (*net.TCPAdd
 //   specified system device (this is primarily
 //   used for Android VpnService routing functionality)
 func (p *Protector) Dial(network, addr string, timeout time.Duration) (net.Conn, error) {
-	op := ops.Begin("protected-dial").Set("addr", addr).Set("timeout", timeout.Seconds())
+	ctx, _ := context.WithTimeout(context.Background(), timeout)
+	return p.DialContext(ctx, network, addr)
+}
+
+// DialContext is same as Dial, but accepts a context instead of timeout value.
+func (p *Protector) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	op := ops.Begin("protected-dial").Set("addr", addr)
+	dl, ok := ctx.Deadline()
+	if ok {
+		op.Set("timeout", dl.Sub(time.Now()).Seconds())
+	}
 	defer op.End()
-	conn, err := p.dial(op, network, addr, timeout)
-	return conn, op.FailIf(err)
+
+	// Dial in goroutine to support arbitrary cancellation.
+	var conn net.Conn
+	var err error
+	chDone := make(chan bool)
+	go func() {
+		conn, err = p.dialContext(op, ctx, network, addr)
+		chDone <- true
```

Comment:
Might be cleaner just to pass the conn itself on the channel. I guess waiting for chDone synchronizes so that we don't end up with concurrent access errors on conn, but just sending conn through the channel is clearer.